File: kraken/pytorch/optim.py
# coding=utf-8

from typing import Dict
import kraken_native


class Optim(object):

  def type(self) -> kraken_native.OptimType:
    raise NotImplementedError

  def conf(self) -> Dict[str, str]:
    raise NotImplementedError


class Adagrad(Optim):

  def __init__(self, weight_decay=None, eps: float = 1e-10):
    super(Adagrad, self).__init__()

    self._weight_decay = weight_decay
    self._eps = eps

  def type(self) -> kraken_native.OptimType:
    return kraken_native.OptimType.kAdagrad

  def conf(self) -> Dict[str, str]:
    conf = dict()

    if self._weight_decay:
      conf['weight_decay'] = str(self._weight_decay)

    conf['eps'] = str(self._eps)

    return conf


class Adam(Optim):

  def __init__(self,
               weight_decay=None,
               beta1: float = 0.9,
               beta2: float = 0.999,
               eps: float = 1e-08,
               amsgrad: bool = False):
    super(Adam, self).__init__()

    self._beta1 = beta1
    self._beta2 = beta2
    self._eps = eps
    self._weight_decay = weight_decay
    self._amsgrad = amsgrad

  def type(self) -> kraken_native.OptimType:
    return kraken_native.OptimType.kAdam

  def conf(self) -> Dict[str, str]:
    conf = dict()
    if self._weight_decay:
      conf['weight_decay'] = str(self._weight_decay)

    conf['beta1'] = str(self._beta1)
    conf['beta2'] = str(self._beta2)
    conf['eps'] = str(self._eps)

    if self._amsgrad:
      conf['amsgrad'] = 'true'
    else:
      conf['amsgrad'] = 'false'

    return conf


class RMSprop(Optim):

  def __init__(self, weight_decay=None, momentum=None, alpha: float = 0.99, eps: float = 1e-8, centered: bool = False):
    super(RMSprop, self).__init__()

    self._weight_decay = weight_decay
    self._momentum = momentum
    self._alpha = alpha
    self._eps = eps
    self._centered = centered

  def type(self) -> kraken_native.OptimType:
    return kraken_native.OptimType.kRMSprop

  def conf(self) -> Dict[str, str]:
    conf = dict()

    if self._weight_decay:
      conf['weight_decay'] = str(self._weight_decay)
    if self._momentum:
      conf['momentum'] = str(self._momentum)

    conf['alpha'] = str(self._alpha)
    conf['eps'] = str(self._eps)

    if self._centered:
      conf['centered'] = 'true'
    else:
      conf['centered'] = 'false'

    return conf


class SGD(Optim):

  def __init__(self, momentum=None, dampening=None, weight_decay=None, nesterov: bool = False):
    super(SGD, self).__init__()

    self._momentum = momentum
    self._dampening = dampening
    self._weight_decay = weight_decay
    self._nesterov = nesterov

  def type(self) -> kraken_native.OptimType:
    return kraken_native.OptimType.kSGD

  def conf(self) -> Dict[str, str]:
    conf = dict()

    if self._momentum:
      conf['momentum'] = str(self._momentum)
    if self._dampening:
      conf['dampening'] = str(self._dampening)
    if self._weight_decay:
      conf['weight_decay'] = str(self._weight_decay)
    if self._nesterov:
      conf['nesterov'] = 'true'
    else:
      conf['nesterov'] = 'false'

    return conf

File: kraken/ps/optim/optim.cc
#include "ps/optim/optim.h"

#include "ps/optim/adagrad.h"
#include "ps/optim/adam.h"
#include "ps/optim/rmsprop.h"
#include "ps/optim/sgd.h"

namespace kraken {

Optim::Optim(OptimType optim_type) : optim_type_(optim_type) {
}

std::unique_ptr<Optim> Optim::Create(
    OptimType optim_type,
    const std::unordered_map<std::string, std::string>& optim_conf) {
  std::unique_ptr<Optim> optim;

  if (optim_type == OptimType::kAdagrad) {
    bool has_weight_decay = false;
    float weight_decay = 0.0;
    float eps = 1e-10;

    if (utils::ParseConf<float>(optim_conf, "weight_decay", &weight_decay)) {
      has_weight_decay = true;
    }

    utils::ParseConf<float>(optim_conf, "eps", &eps);

    optim.reset(new Adagrad(has_weight_decay, weight_decay, eps));
  } else if (optim_type == OptimType::kAdam) {
    bool has_weight_decay = false;
    float weight_decay = 0.0;
    float beta1 = 0.9;
    float beta2 = 0.999;
    float eps = 1e-08;
    bool amsgrad = false;

    if (utils::ParseConf<float>(optim_conf, "weight_decay", &weight_decay)) {
      has_weight_decay = true;
    }

    utils::ParseConf<float>(optim_conf, "beta1", &beta1);
    utils::ParseConf<float>(optim_conf, "beta2", &beta2);
    utils::ParseConf<float>(optim_conf, "eps", &eps);
    utils::ParseConf<bool>(optim_conf, "amsgrad", &amsgrad);

    optim.reset(
        new Adam(has_weight_decay, weight_decay, beta1, beta2, eps, amsgrad));
  } else if (optim_type == OptimType::kRMSprop) {
    bool has_weight_decay = false;
    float weight_decay = 0.0;
    bool has_momentum = false;
    float momentum = 0.0;
    float alpha = 0.99;
    float eps = 1e-8;
    bool centered = false;

    if (utils::ParseConf<float>(optim_conf, "weight_decay", &weight_decay)) {
      has_weight_decay = true;
    }

    if (utils::ParseConf<float>(optim_conf, "momentum", &momentum)) {
      has_momentum = true;
    }

    utils::ParseConf<float>(optim_conf, "alpha", &alpha);
    utils::ParseConf<float>(optim_conf, "eps", &eps);
    utils::ParseConf<bool>(optim_conf, "centered", &centered);

    optim.reset(new RMSprop(has_weight_decay, weight_decay, has_momentum,
                            momentum, alpha, eps, centered));
  } else if (optim_type == OptimType::kSGD) {
    bool has_weight_decay = false;
    float weight_decay = 0.0;
    bool has_momentum = false;
    float momentum = 0.0;
    bool has_dampening = false;
    float dampening = 0.0;
    bool nesterov = false;

    if (utils::ParseConf<float>(optim_conf, "weight_decay", &weight_decay)) {
      has_weight_decay = true;
    }

    if (utils::ParseConf<float>(optim_conf, "momentum", &momentum)) {
      has_momentum = true;
    }

    if (utils::ParseConf<float>(optim_conf, "dampening", &dampening)) {
      has_dampening = true;
    }

    utils::ParseConf<bool>(optim_conf, "nesterov", &nesterov);

    optim.reset(new SGD(has_weight_decay, weight_decay, has_momentum, momentum,
                        has_dampening, dampening, nesterov));
  }

  return optim;
}

}  // namespace kraken
